File: zerOS/src/kernel/cpu/misc.c
#include <kernel/cpu/misc.h>
#include <kernel/cpu/io.h>

#include <misc/sections.h>

extern void zerOS_halt(void)
{
    asm volatile(
        "hlt" : : : "memory"
    );
}

extern void zerOS_reboot(void)
{
    zerOS_outw(0x64, 0xFE | (1 << 8));
    zerOS_halt();
}

extern void zerOS_cli(void)
{
    asm volatile(
        "cli" : : : "memory"
    );
}

extern void zerOS_hcf(void)
{
    zerOS_cli();
    while (true)
    {
        zerOS_halt();
    }
}

File: zerOS/src/kernel/cpu/misc.rs
use core::arch::asm;

use super::irq;

use crate::kernel::cpu::io::outw;

pub fn halt()
{
    unsafe {
        asm! {
            "hlt",
            options(att_syntax),
            options(nomem),
            options(nostack)
        }
    }
}

pub fn reboot() -> !
{
    outw(0x64, 0xFE | (1 << 8));
    halt();
    unreachable!();
}

/// Halt and catch fire
pub fn hcf() -> !
{
    loop
    {
        irq::disable();
        halt();
    }
}
